a closure called instantly with an `or` block fails to compile

### Describe the bug

Trying to inline a function with an `Option` return which is invoked immediately  results in a compilation error when using an `or` block.

### Reproduction Steps

The following code results in the described bug behavior:
```V
server_script_path := fn [tmp_wd] () !string {
    server_script_embedded_file := $embed_file('./src/server.sh')
    path := os.join_path(tmp_wd, 'server.sh')
    os.write_file(path, server_script_embedded_file.to_string())!
    return path
}() or { panic(err) }
```

### Expected Behavior

I would expect this to compile without issue.

### Current Behavior

Does not compile:
```
/tmp/v_501/core_v.01HTDW22W7WKSQ5Q8XPN9GNMEG.tmp.c:23570:9: error: initializing 'string' (aka 'struct string') with an expression of incompatible type '_result_string (*)()' (aka 'struct _result_string (*)()')
        string server_script_path =
               ^
/tmp/v_501/core_v.01HTDW22W7WKSQ5Q8XPN9GNMEG.tmp.c:23571:5: error: expected ';' at end of declaration
        _t5     _result_string _t6 = __closure_create(anon_fn_143c1662f75e5a6c___result_string_1704, (struct _V_anon_fn_143c1662f75e5a6c___result_string_1704_Ctx*) memdup_uncollectable(&(struct _V_anon_fn_143c1662f75e5a6c___result_string_1704_Ctx){.tmp_wd = tmp_wd,
           ^
           ;
/tmp/v_501/core_v.01HTDW22W7WKSQ5Q8XPN9GNMEG.tmp.c:23572:74: error: called object type 'void *' is not a function or function pointer
        }, sizeof(struct _V_anon_fn_143c1662f75e5a6c___result_string_1704_Ctx)))();
```

The following does however compiler:
```V
server_script_path := fn [tmp_wd] () string {
	server_script_embedded_file := $embed_file('./src/server.sh')
	path := os.join_path(tmp_wd, 'server.sh')
	os.write_file(path, server_script_embedded_file.to_string()) or { panic(err) }
	return path
}()
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

0.4.5 e6b43a1

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.5 365bd18.e6b43a1
OS: macos, macOS, 14.4, 23E214
Processor: 10 cpus, 64bit, little endian, Apple M1 Max

getwd: /X/Y/Z
vexe: /X/Y/Z/v/v
vexe mtime: 2024-03-20 20:08:31

vroot: OK, value: /Users/X/DevTools/v
VMODULES: OK, value: /Users/X/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.44.0
Git vroot status: weekly.2024.12-13-ge6b43a16 (82 commit(s) behind V master)
.git/config present: true

CC version: Apple clang version 15.0.0 (clang-1500.3.9.4)
thirdparty/tcc status: thirdparty-macos-arm64 5c1d002f
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -790,21 +790,25 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 	}
 	// NOTE: everything could be done this way
 	// see my comment in parser near anon_fn
+	mut tmp_anon_fn_var := ''
 	if node.left is ast.AnonFn {
 		if node.left.inherited_vars.len > 0 {
-			tmp_var := g.new_tmp_var()
+			tmp_anon_fn_var = g.new_tmp_var()
 			fn_type := g.fn_var_signature(node.left.decl.return_type, node.left.decl.params.map(it.typ),
-				tmp_var)
+				tmp_anon_fn_var)
 			line := g.go_before_last_stmt().trim_space()
 			g.empty_line = true
 			g.write('${fn_type} = ')
 			g.expr(node.left)
 			g.writeln(';')
+			g.set_current_pos_as_last_stmt_pos()
 			g.write(line)
-			if g.out.last_n(1) != '\n' {
-				g.writeln('')
+			if node.or_block.kind == .absent {
+				if g.out.last_n(1) != '\n' {
+					g.writeln('')
+				}
+				g.write(tmp_anon_fn_var)
 			}
-			g.write(tmp_var)
 		} else if node.or_block.kind == .absent {
 			g.expr(node.left)
 		}
@@ -887,7 +891,11 @@ fn (mut g Gen) call_expr(node ast.CallExpr) {
 			}
 			g.write('${styp} ${tmp_opt} = ')
 			if node.left is ast.AnonFn {
-				g.expr(node.left)
+				if node.left.inherited_vars.len > 0 {
+					g.write(tmp_anon_fn_var)
+				} else {
+					g.expr(node.left)
+				}
 			}
 		}
 	}

--- a/vlib/v/tests/closure_option_direct_call_test.v
+++ b/vlib/v/tests/closure_option_direct_call_test.v
@@ -1,0 +1,8 @@
+fn test_closure_option_direct_call() {
+	tmp_wd := 'aaa'
+	b := fn [tmp_wd] () !string {
+		return tmp_wd
+	}() or { panic(err) }
+	println(b)
+	assert b == 'aaa'
+}

